### ISSUE TITLE
not tracking the WCF service calls from a web api

### ISSUE THREAD

**Issue by vemulakiran** [Microsoft/ApplicationInsights-dotnet-server#302]
My web api is invoking a wcf service method and it is  not being tracked as a Dependency call in Applicaiton Insight. I can see sql dependencies, but not the wcf service. 

Let me what changes I need to track the WCF service calls.

**Comment by SergeyKanzhelev** [comment]
What transport do you for your wcf service? Application Insights will only collect http transport out of the box. @tomasr may have ideas. He is an author of https://github.com/Microsoft/ApplicationInsights-SDK-Labs/tree/master/WCF

**Comment by tomasr** [comment]
Yes, even with the default HTTP dependency tracking, it will be a bit lacking, as it will only tell you how long the call lasted and the service URL; it won't tell you what operation you're calling on the service.

I've considered for a bit implementing a library that could be added to WCF endpoint clients to add the necessary `TrackDependency()` calls as an alternative (you can certainly do that yourself).

**Comment by vemulakiran** [comment]
I am using netTcpBinding to consume the WCF service.  what changes I need to do in the Web API to track the dependency calls over netTCP.

**Comment by tomasr** [comment]
@vemulakiran The simplest, but most intrusive way would be to wrap your calls with `TelemetryClient.TrackDependency()` calls to emit the necessary dependency telemetry events.

A less invasive alternative would be to write a reusable, custom `IClientMessageInspector` or using another similar WCF extensibility point that generates the dependency telemetry events, and then inject it into the client runtime using an `EndpointBehavior`. This still requires manually registering the behavior for each single client, unfortunately.

@SergeyKanzhelev FWIW, I've been investigating other means of automatically generating the telemetry events:

* I've thought about implementing that `EndpointBehavior` myself, but there's no way to "install a NuGet Package" and have it work right away.
* WCF does have some great ETW events that would provide *most* of the necessary info on 4.5+, but the ETW implementation is not based on `EventSource`, so it's not observable in-proc using `EventListener`.
* Extending the current profiler-based model would work (I've prototyped it far enough to know it would be possible, at least for basic scenarios), but not sure if that's a road we'd like to look into.

**Comment by SergeyKanzhelev** [comment]
@tomasr thanks for investigation!

I think the `EndpointBehavior`  approach is very promising as it is least invasive and easily configurable by customer. We most probably will have a similar "connection factory" approach for RedisCache, Signal-R and other dependencies collection. 


Extending of the profiler is not a bad option either. Will we need to instrument .NET framework methods only or also customer's code? Will it be possible to get an operation name using profiler?

**Comment by tomasr** [comment]
Re: Profiler: what I've tried so far is instrumenting framework code: https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ServiceChannel.cs#L720, but that only covers sync calls, so still would need to add the async path.

That said, I haven't found an easy way to get the raw operation name, but the SOAP Action, which is just as good, I think, for most purposes. It's very likely that doing the `EndpointBehavior` path would yield the same limitation, in any case.

The tricky part would be dealing with the default HTTP dependency tracking in HTTP-based operations....

**Comment by vemulakiran** [comment]
@tomasr @SergeyKanzhelev  Thanks for the useful information

I am able to track the net.tcp dependency call from AfterReceiveReply method in IClientMessageInspector implementation, but the dependency track events are not correlated with current web api request and other dependency call events of the request. I presume this is because  I am using Async methods. Also, I noticed that dependency events are tracked in case of network related exception If I use IClientMessageInspector

So, I have decided to wrap the wcf calls with TelemetryClient.TrackDependency() calls and it is working fine.

**Comment by tomasr** [comment]
@vemulakiran Thanks for the information. Out of curiosity, where you using the `BeginXxx()` methods, or using `Task`-based calls in your async scenario?

**Comment by vemulakiran** [comment]
@tomasr  I am using async calls with await operators (task based).

**Comment by tomasr** [comment]
FWIW, I've continued investigating this in a bit more detail, as it's quite fun. Here are some findings:

* Hooking `ServiceChannel.Call()` through the profile APIs is suboptimal because it would only work with the profiler. Using a service behavior is better (as it would be usable if the profiler is absent).

* Using the regular WCF extensibility points is not an option due to the issues already mentioned. What does work, however, is hooking directly into the WCF channel stack so that one can trap all the channel open and message operations.

* This is, unfortunately, *far* more complex than it appears at first glance. There's a significant amount of work required to support multiple channel shapes and all the async code paths in the channel stack. `NetTcpBinding` is particularly nasty here because it forces the channel stack to be based on `IDuplexChannel` (if you're using the default buffered transfer mode). It does seem to provide the best results I've gotten, so far.

* The issues with async code are real, but I haven't quite figured out *why* they are a problem. From what I can see in the WCF code, the Task-based API merely uses `TaskFactory<TResult>.FromAsync()`. One would expect this to flow the context correctly provided you're on 4.5+ (see http://stackoverflow.com/questions/18383923/why-is-httpcontext-current-null-after-await), but so far, I haven't gotten it to work.

Anyway, here's an example of some data tracked using my proof of concept with profiler-based automatic instrumentation + custom WCF channels using `NetTcpBinding` to invoke the service:
![image](https://cloud.githubusercontent.com/assets/16179/23024256/61677b76-f427-11e6-943e-bb6d46dd1a38.png)

**Comment by mlooise** [comment]
Hi @tomasr . Looks like you made some interesting progress towards what i'm looking for. Could you maybe share your code as a pull-request or did you maybe make your experiments in a feature branch that you could publish. I'm quite interested in the code behind it. I think it could eventueally be a very usefull addon to our 'insights' :smile:

**Comment by SergeyKanzhelev** [comment]
@tomasr, this is a great progress. Do you think we can include at least a basic behavior that will work in most cases as a PR? After we will ensure it's working fine - we can enable it automatically using Profiler.

**Comment by tomasr** [comment]
@SergeyKanzhelev What project would you like this on? Right now I've coded my code as part of the WCF package, but I can split it into whatever you want....

**Comment by SergeyKanzhelev** [comment]
WCF for now is OK. I'd say this one once ready and validated: https://github.com/Microsoft/ApplicationInsights-dotnet-server/tree/develop/Src/DependencyCollector

**Comment by tomasr** [comment]
Initial code is now in https://github.com/tomasr/ApplicationInsights-SDK-Labs/tree/WcfClient, and PR https://github.com/Microsoft/ApplicationInsights-SDK-Labs/pull/94
